Sources/Workspace/Workspace.swift

```diff
@@ -2172,17 +2168,21 @@ extension Workspace {
                 delegateQueue: .sharedConcurrent,
                 callbackQueue: .sharedConcurrent
             ) { result in
+                #warning("cleanup")
+                var result = result
                 switch result {
-                // Diagnostics.fatalError indicates that a more specific diagnostic has already been added.
-                case .failure(Diagnostics.fatalError):
-                    self.delegate?.didLoadManifest(packagePath: packagePath, url: packageLocation, version: version, packageKind: packageKind, manifest: nil, diagnostics: manifestLoadingDiagnostics.get())
                 case .failure(let error):
-                    manifestLoadingScope.emit(error)
-                    self.delegate?.didLoadManifest(packagePath: packagePath, url: packageLocation, version: version, packageKind: packageKind, manifest: nil, diagnostics: manifestLoadingDiagnostics.get())
+                    manifestLoadingDiagnostics.append(.error(error))
+                    self.delegate?.didLoadManifest(packagePath: packagePath, url: packageLocation, version: version, packageKind: packageKind, manifest: nil, diagnostics: manifestLoadingDiagnostics)
                 case .success(let manifest):
-                    manifestLoadingScope.trap { try self.validateManifest(manifest) }
-                    self.delegate?.didLoadManifest(packagePath: packagePath, url: packageLocation, version: version, packageKind: packageKind, manifest: manifest, diagnostics: manifestLoadingDiagnostics.get())
+                    let validationDiagnostics = ManifestValidator.validate(manifest, sourceControlValidator: self.repositoryManager, fileSystem: self.fileSystem)
+                    if !validationDiagnostics.isEmpty {
+                        result = .failure(Diagnostics.fatalError)
+                        manifestLoadingDiagnostics.append(contentsOf: validationDiagnostics)
+                    }
+                    self.delegate?.didLoadManifest(packagePath: packagePath, url: packageLocation, version: version, packageKind: packageKind, manifest: manifest, diagnostics: manifestLoadingDiagnostics)
                 }
+                manifestLoadingScope.emit(manifestLoadingDiagnostics)
```

Comment:
here is the key change (and motivation): it allows us to get rid of the fake observability scope and special casing Diagnostics.fatalError

Reply:
That looks like a good approach.  And this means the `didLoadManifest()` callback and the diagnostics handler will always see all the diagnostics, from the looks of it.

Does that mean that a client that just wanted to capture diagnostics at the end (in `didLoadManifest()`) could pass a no-op manifestLoadingScope and just implement the delegate callback?  In practice it seems that any given client would only want to do one or the other, right, otherwise it would get duplicate diagnostics.

Reply:
I realize this might not be any different from before because of this PR, but it's always been somewhat unclear here what a client should ideally do to avoid duplication.

Reply:
> Does that mean that a client that just wanted to capture diagnostics at the end (in didLoadManifest()) could pass a no-op manifestLoadingScope and just implement the delegate callback? In practice it seems that any given client would only want to do one or the other, right, otherwise it would get duplicate diagnostics.

who is the client in this case? this is private function, and so the observabilityScope passed to it may contain other diagnostics above and beyond manifest loading

Reply:
Sorry, I misread this and didn't realize that it's a private function (thus the client would be the client of libSwiftPM, but that point is moot, as it's private).

I agree with the other sentiment and think we'll want to revisit the delegate here at some point.  There's a bit of a split personality between the delegate callbacks but then the diagnostics in the scope.